torchdata/datapipes/iter/util/cacheholder.py

```diff
@@ -67,71 +68,197 @@ def __len__(self) -> int:
+        self._end_caching_flag = True
+        # Re-assign source_datapipe
+        self.source_datapipe = todo_dp
+
+        # Cached: keep filepath(s)
+        cached_dp = cached_dp.map(fn=self.filepath_fn)
+        # Convert list back to single elements
+        return cached_dp.unbatch(-1)
+
+
+def _read_bytes(fd):
+    return b"".join(fd)
+
+
+def _read_str(fd):
+    return "".join(fd)
+
+
+@functional_datapipe("end_caching")
+class EndOnDiskCacheHolderIterDataPipe(IterDataPipe):
+    """
+    `EndOnDiskCacheHolder` is a IterDataPipe that indicates when the result of
+    prior DataPipe will be saved local files specified by `filepath_fn`
+    And, the result of source DataPipe is required to be a tuple of metadata and data,
+    or a tuple of metadata and file handle.
+
+    Args:
+        datapipe: IterDataPipe with at least one `OnDiskCacheHolder` in the graph.
+        mode: Mode in which cached files are opened for write the data. This is needed
+            to be aligned with the type of data or file handle from `datapipe`.
+        filepath_fn: Optional function to extract filepath from the metadata from `datapipe`.
+            As default, it would directly use the metadata as file path.
+        same_filepath_fn: Set to `True` to use same `filepath_fn` from the `OnDiskCacheHolder`.
+        skip_read: Boolean value to skip reading the file handle from `datapipe`.
+            As default, reading is enabled and reading function is created based on the `mode`.
+    """
+
+    def __new__(cls, datapipe, mode="w", filepath_fn=None, *, same_filepath_fn=False, skip_read=False):
+        if filepath_fn is not None and same_filepath_fn:
+            raise ValueError("`filepath_fn` is mutually exclusive with `same_filepath_fn`")
+
+        graph = traverse(datapipe, exclude_primitive=True)
+        # Get the last CacheHolder
+        cache_holder = EndOnDiskCacheHolderIterDataPipe._recursive_search(graph)
+        if cache_holder is None:
+            raise RuntimeError("Expected `OnDiskCacheHolder` existing in pipeline when `end_caching` is invoked")
+        if cache_holder._end_caching_flag:
+            raise RuntimeError("`end_caching` can only be invoked once per `OnDiskCacheHolder`")
+
+        cached_dp = cache_holder._end_caching()
+
+        if same_filepath_fn:
+            filepath_fn = cache_holder.filepath_fn
+
+        todo_dp = datapipe
+        if not skip_read:
+            if "b" in mode:
+                todo_dp = todo_dp.map(fn=_read_bytes, input_col=1)
+            else:
+                todo_dp = todo_dp.map(fn=_read_str, input_col=1)
+
+        todo_dp = todo_dp.map(fn=filepath_fn, input_col=0)
+
+        # Extra hash check here when hash is provided.
+        # And, raise Error if data returned from prior operations doesn't meet hash
+        if cache_holder.hash_dict is not None:
+            todo_dp = todo_dp.check_hash(cache_holder.hash_dict, cache_holder.hash_type)
```

Comment:
This part is added based on the feedback from TorchText. If `hash_dict` is specified, after data is downloaded, we need to check the hash again just in case the data source has incorrect data.